src/lib.rs

```diff
@@ -482,8 +484,19 @@ where
         match self.key_extractor.extract(&req) {
             // Extraction worked, let's check if rate limiting is needed.
             Ok(key) => match self.limiter.check_key(&key) {
-                Ok(_) => {
+                Ok(_positive) => {
                     let fut = self.service.call(req);
+                    // let mut response = ...;
+                    // // Add x-ratelimit headers
+                    // response.headers_mut().insert(
+                    //     HeaderName::from_static("x-ratelimit-limit"),
+                    //     positive.quota().burst_size().get().into(),
+                    // );
+                    // response.headers_mut().insert(
+                    //     HeaderName::from_static("x-ratelimit-remaining"),
+                    //     positive.remaining_burst_capacity().into(),
+                    // );
```

Comment:
Can't you use something like this?

```rust
fut.headers_mut().insert(
     HeaderName::from_static("x-ratelimit-limit"),
    quota().burst_size().get().into(),
);
```

The future (fut) should have the type `ServiceResponse`, but I might be wrong.

Reply:
I only have pool or redy_to_pool, something like this appears in the autocomplit, I will check for it

Reply:
> Can't you use something like this?

No i can't

![Screenshot_select-area_20220424224207](https://user-images.githubusercontent.com/59842932/164993736-d434bbf3-2cb8-4530-a8b2-cb4350a52dfd.png)

Reply:
Yeah, I've overlooked that [Service::call](https://docs.rs/actix-web/latest/actix_web/dev/trait.Service.html#tymethod.call) does return `Service::Future` not `Service::Result`. Otherwise this would have been rather easy. I this case I think we have to implement a custom future. This will likely be a rougher experience since actix-web and async Rust are not the easiest things to work with. But I will help you as much as I can.

Since futures are basically state machines, my idea would be to implement a future that just polls the underlying response until it's ready and then inserts the headers.

Don't hesitate to ask further questions. Async Rust at this level is complex, I'm not even entirely sure if my suggestion will work. If you're not already familiar with futures I recommend reading the tokio [tutorial](https://tokio.rs/tokio/tutorial).

Reply:
Look at this example
https://github.com/chefsale/actix-ratelimit/blob/master/src/middleware.rs#L219-L236

Do I do the same?
I will have to change the Future to
https://github.com/chefsale/actix-ratelimit/blob/master/src/middleware.rs#L153

Reply:
> Do I do the same?

Actually I tried, but I got a Lifetime Error, I will try to solve it and if I can't I will ask for your help 🙂

Reply:
That's already a good orientation and yes, you need to change the future type in the `Service` implementation. `Box::pin` comes with a slight performance cost and potentially it might be easier to implement our own future type here (also regarding lifetimes). In general you can simply create a struct that stores the rate limiter data and the response future. And then you only need to implement `Future` for that type. At least it sounds good to me in theory.